Allow extension override.

MancunianSam opened the pull request:
TDR have the case where there files in S3 have no extension. This allows
you to provide an extension which DROID will use rather than deriving it
from the name.

techncl reviewed (commented):
Looks good! Just some minor comments. Also, I was wondering, if someone accidentally provided an extension even though the file has one and the extension was incorrect i.e. didn't match the one in the file, what would happen and should we check for that?

MancunianSam commented on droid-core-interfaces/src/main/java/uk/gov/nationalarchives/droid/core/interfaces/resource/S3IdentificationRequest.java:
```diff
@@ -84,7 +85,15 @@ public final void open(final S3Uri theFile) throws IOException {
     @Override
     public final String getExtension() {
-        return ResourceUtils.getExtension(requestMetaData.getName());
+        return this.extension != null ? this.extension : ResourceUtils.getExtension(requestMetaData.getName());
```
1. Yes
2. Yes
3. No, I don't want to do too much validation in case there's some weird edge case it doesn't catch. 
4. See 3

MancunianSam commented on droid-core-interfaces/src/main/java/uk/gov/nationalarchives/droid/core/interfaces/resource/HttpIdentificationRequest.java:
```diff
@@ -83,7 +84,15 @@ public final void open(final URI theFile) throws IOException {
     @Override
     public final String getExtension() {
-        return ResourceUtils.getExtension(requestMetaData.getName());
+        return extension == null ? ResourceUtils.getExtension(requestMetaData.getName()) : extension;
```
As above.

MancunianSam commented on droid-core-interfaces/src/main/java/uk/gov/nationalarchives/droid/core/interfaces/resource/FileSystemIdentificationRequest.java:
```diff
@@ -92,7 +91,15 @@ public final void open(final Path theFile) throws IOException {
     @Override
     public final String getExtension() {
-        return extension;
+        return extension == null ? ResourceUtils.getExtension(fileName) : extension;
```
As above.

MancunianSam commented on droid-core-interfaces/src/main/java/uk/gov/nationalarchives/droid/core/interfaces/resource/HttpIdentificationRequest.java:
```diff
@@ -83,7 +84,15 @@ public final void open(final URI theFile) throws IOException {
+    }
+
+    /**
+     * Sets the file extension. This is used when it can't be determined from the file name
```
I see your point, it isn't quite right. I've changed it to `Sets the file extension. If this is set, this will be used as the file extension instead of an extension derived from the name`

sparkhi reviewed (commented):
Looks good to me. One question, perhaps from user expectation point of view.

If they provide a "docx" file and extension as "pdf" - we need to return `ExtensionMismatch` as `true` in the result. 

It may be a good idea to add a small test which uses `word97` in the resources but runs by passing in explicit extension as "pdf" or something

MancunianSam commented:
I've added a test which uses the word97 test file and passed pdf as the extension and we do get a container signature match and an extension mismatch. 

While I was there, I found a bug in the HttpWindowReader which I've fixed in the same commit.
